pkg/ipcache/metadata.go

```diff
@@ -427,18 +427,28 @@ func (ipc *IPCache) resolveIdentity(ctx context.Context, prefix netip.Prefix, in
 	}
 
 	lbls := info.ToLabels()
-	if (lbls.Has(labels.LabelWorld[labels.IDNameWorld]) ||
-		lbls.Has(labels.LabelWorldIPv4[labels.IDNameWorldIPv4]) ||
-		lbls.Has(labels.LabelWorldIPv6[labels.IDNameWorldIPv6])) &&
-		(lbls.Has(labels.LabelRemoteNode[labels.IDNameRemoteNode]) ||
-			lbls.Has(labels.LabelHost[labels.IDNameHost])) {
-		// If the prefix is associated with both world and (remote-node or
-		// host), then the latter (remote-node or host) take precedence to
-		// avoid allocating a CIDR identity for an entity within the cluster.
+
+	// If we are restoring an identity in the remote-node identity scope, we need to unconditionally add remote-node and cidr labels back
+	if restoredIdentity.Scope() == identity.IdentityScopeRemoteNode {
+		lbls.MergeLabels(labels.LabelRemoteNode)
+		cidrLabels := cidrlabels.GetCIDRLabels(prefix)
+		lbls.MergeLabels(cidrLabels)
+	}
+
+	// If the prefix is associated with the host or remote-node, then
+	// force-remove the world label.
+	if lbls.Has(labels.LabelRemoteNode[labels.IDNameRemoteNode]) ||
+		lbls.Has(labels.LabelHost[labels.IDNameHost]) {
 		n := lbls.Remove(labels.LabelWorld)
 		n = n.Remove(labels.LabelWorldIPv4)
 		n = n.Remove(labels.LabelWorldIPv6)
-		n = n.Remove(cidrlabels.GetCIDRLabels(prefix))
+
+		// It is not allowed for nodes to have CIDR labels, unless policy-cidr-selects-nodes
+		// is enabled. Then CIDR labels are permitted.
+		// So, remove if necessary
+		if !option.Config.PolicyCIDRSelectsNodes {
+			n = n.Remove(cidrlabels.GetCIDRLabels(prefix))
+		}
```

Comment:
This doesn't seem necessary considering my previous comment in this file

Reply:
LGTM now.